src/Npgsql/TypeHandlers/HstoreHandler.cs

```diff
@@ -181,6 +183,29 @@ async ValueTask<T> ReadInto<T>(T dictionary, int numElements, NpgsqlReadBuffer b
         }
 
         #endregion
+
+        #region IReadOnlyDictionary
+
+        /// <inheritdoc />
+        public int ValidateAndGetLength(
+            IReadOnlyDictionary<string, string?> value, ref NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter)
+            => ValidateAndGetLength((IDictionary<string, string?>)value, ref lengthCache, parameter);
+
+        /// <inheritdoc />
+        public Task Write(IReadOnlyDictionary<string, string?> value,
+            NpgsqlWriteBuffer buf, NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter, bool async)
+            => Write((IDictionary<string, string?>)value, buf, lengthCache, parameter, async);
```

Comment:
This is a bug - there's no guarantee that the value we're passing here is an `IDictionary<string, string?>` so the cast might be invalid. How to fix this depends on how `IReadOnlyDictionary` is handled across frameworks.

Reply:
Just made a private method to write a `IReadOnlyDictionary<,>` and redirect writes from other places to it. For `IDictionary<,>` will be another method, but with the same body.

Reply:
I've added a private method that takes an `IEnumerable<KeyValuePair<string, string?>>` and an explicit count to avoid duplicating the method, which I think is a bit cleaner - happy to just duplicate it if you prefer though.

Reply:
Yep, that's better (: